pubtools/_content_gateway/push_base.py

```diff
@@ -0,0 +1,463 @@
+
+                Intended to be called as the first operation in
+                the content gateway workflow.
+        """
+
+        all_products = self.cgw_client.get_products()
+        for product in all_products:
+            yield (product['name'], product['productCode']), product['id']
+
+    def _fetch_product_version(self, product_name, product_code, version_name):
+        """
+        Get the list of product versions from the content gateway
+        by calling its client methods.
+
+        Args:
+            product_name (str):
+                product name of the product
+            product_code (str):
+                product code of the product
+            version_name (str):
+                version name of the product
+
+        Returns:
+            Yields (dict):
+                Returns the version records associate with the product.
+                The returned structure is an iterator to reduce memory requirements.
+        """
+
+        product_id = self.product_mapping.get((product_name, product_code))
+        all_versions = self.cgw_client.get_versions(product_id)
+        for version in all_versions:
+            yield (product_name, product_code, version['versionName']), version['id']
+
+    def _fetch_file(self, product_name, product_code, version_name, download_url):
+        """
+        Get the list of file records associate with version and product
+        from the content gateway by calling its client methods.
+
+        Args:
+            product_name (str):
+                product name of the file
+            product_code (str):
+                product code of the file
+            version_name (str):
+                version name of the file
+            download_url (str):
+                download url of the file
+
+        Returns:
+            Yields (dict):
+                Returns the file records.
+                The returned structure is an iterator to reduce memory requirements.
+        """
+
+        product_id = self.product_mapping.get((product_name, product_code))
+        version_id = self.pv_mapping.get((product_name, product_code, version_name))
+        all_files = self.cgw_client.get_files(product_id, version_id)
+        for data in all_files:
+            yield (product_name, product_code, version_name, data['downloadURL']), data.get('id')
+
+    def __init__(self, cgw_hostname, cgw_username, cgw_password):
+        """
+        Initialize.
+
+        Args:
+            cgw_hostname (str):
+                CGW registry URL
+            cgw_username (str):
+                username for CGW HTTP API
+            cgw_password (str):
+                password for CGW HTTP API
+        """
+
+        self.auth = CGWBasicAuth(cgw_username, cgw_password)
+        self.cgw_client = CGWClient(cgw_hostname, self.auth)
+        self.product_mapping = FetcherDict({}, fetcher=self._fetch_product, key_checker=self._product_mapping_key_check)
```

Comment:
Val arg is not necessary here

Reply:
The val will be keep updating as we add products and related versions. Initially, it will be empty.